### PR TITLE
opensuse: Use dnf instead of zypper

### PR DESCRIPTION
**Pull request by DaanDeMeyer**
Let's reduce the number of package managers we use by using dnf to
build opensuse images.

The only caveat is that extra GPG keys are listed in repomd.xml for
the opensuse repos, so we have to download that file and parse it to
figure out the extra keys that we have to write to the repo configs.

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
> I'm not sure I like this, I don't even know if there's a working and reliable dnf implementation when doing native builds. And if there is and it doesn't work, it's going to be unsupported and tricky to fix

I've been in contact with someone from opensuse with regards to some question about the gpg keys, and he mentioned that there's quite a few people using dnf with opensuse. 

It even has a wiki page and everything: https://en.opensuse.org/SDB:DNF

File: mkosi/distributions/fedora.py [hunk 1]
```diff
@@ -90,39 +89,33 @@ def url_exists(url: str) -> bool:
...
                 )
             )
 
+            ident = 0
+            for url in repo.gpgurls:
+                f.write(ident * " ")
+                f.write(f"{url}\n")
+                ident = len("gpgkey=")
+
+            f.write("\n")
```
[the remark this answers is not in the document]
Comment (DaanDeMeyer) [reply mkosi/distributions/fedora.py]: Did it slightly different

**Comment by DaanDeMeyer** [comment]
> But is it fully supported? I mean, can I rely on it when building an opensuse image on opensuse on OBS?

I wouldn't see why not? dnf is packaged in opensuse, and it's all just rpm under the hood.

**Comment by DaanDeMeyer** [comment]
What can happen is that they decide to do one of these weird hacks like putting extra gpgkeys in repomd.xml and we'd have to add some extra logic to make that work with dnf likely. But given that it's literally the only thing we had to do to make dnf work I'd say the risk of that is pretty low.

File: mkosi/distributions/fedora.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (behrmann) [reply mkosi/distributions/fedora.py]: That's also a good solution, though a comment will probably be clearer then recalculating the indentation every time
```suggestion
                # indent subquent keys to len("gpgkey=")
                f.write("gpgkey=" if not i else "       ")
```

File: mkosi/distributions/fedora.py [hunk 2]
```diff
@@ -145,6 +145,11 @@ def invoke_dnf(
         "--noplugins",
     ]
 
+    # Make sure we download filelists so all dependencies can be resolved.
+    # See https://bugzilla.redhat.com/show_bug.cgi?id=2180842
```
Comment (keszybz) [inline mkosi/distributions/fedora.py:149]: That bug seems quite confused. Did you encounter some issue with mkosi or is this just a precaution?

File: mkosi/distributions/opensuse.py [hunk 3]
```diff
@@ -1,13 +1,13 @@
 # SPDX-License-Identifier: LGPL-2.1+
 
+import urllib.request
+import xml.etree.ElementTree as ET
```
Comment (keszybz) [inline mkosi/distributions/opensuse.py:4]: `import xml.etree.ElementTree as ElementTree` ? It is used exactly once, so this is just obfuscation.

File: mkosi/distributions/opensuse.py [hunk 4]
```diff
@@ -34,74 +34,38 @@ def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bo
...
+def fetch_gpgurls(repourl: str) -> list[str]:
+    gpgurls = [f"{repourl}/repodata/repomd.xml.key"]
+
+    with urllib.request.urlopen(f"{repourl}/repodata/repomd.xml") as f:
+        xml = f.read().decode()
+        root = ET.fromstring(xml)
+
+        tags = root.find("{http://linux.duke.edu/metadata/repo}tags")
+        if not tags:
+            die("repomd.xml missing tags element")
```
Comment (keszybz) [inline mkosi/distributions/opensuse.py:64]: `<tags>`?

File: mkosi/distributions/fedora.py (hunk 2 above)
Comment (keszybz) [reply mkosi/distributions/fedora.py:149]: Yeah, I think this patch can be dropped. After the latest updates, it's just a bunch of fringe packages. `dnf5` is not even installed on older releases, so we don't have to worry so much about it.

Comment (DaanDeMeyer) [reply mkosi/distributions/fedora.py:149]: @keszybz Does that include all the centos packages? I still need to build centos 8 with mkosi.

Comment (keszybz) [reply mkosi/distributions/fedora.py:149]: Ah, OK. I didn't think about the case where you use `dnf5` from the host to install an older distro. So let's merge this, and then maybe we can skip it again on Fedora >= 38. But that can be done later.

**Pull request merged**